xds/src/main/java/io/grpc/xds/internal/sds/ServerWrapperForXds.java

```diff
@@ -56,26 +59,28 @@
   }
 
   @Override
-  public Server start() throws IOException {
+  public synchronized Server start() throws IOException {
```

Comment:
We can use an `AtomicBoolean` and checkState `compareAndSet(false, true)` instead. `synchronized` method with IO operation scares me in general although it works in this case.

Reply:
I do understand your concern about IO operation inside `synchronized` and thought about it. Note the lock is not for just atomic setting of `started` but for an atomic start operation similar to how [`ServerImpl.start`](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ServerImpl.java#L175) does it. There too IO operations are done inside the synchronized block. If 2 threads concurrently call `start`, one of them will fail with "Already started" (assuming the other `start` succeeds) in this implementation and I thought that was one of the objectives. 

So the alternatives are:
- don't worry about concurrent start protection and just implement protection against repetitive (non-concurrent) calls to `start` using just the `started` as an `AtomicBoolean`
- do not worry about side-effects and let the `delegate.start()` throw the exception with "Already started" (old code)

WDYT?

Reply:
Oh, I looked at the documentation, it says `IllegalStateException - if already started`. I guess start() didn't do very much work before, so it would run reasonably quickly. Now we are making it take a much longer time. It seems quite fair to use an AtomicBoolean and say "already started" even before the other start() returns. (I was mistaken and thought a duplicate start would just return immediately.)

Reply:
> ... (Although multiple simultaneous starts really isn't all that interesting of a use-case.)

I agree with that but `ServerImpl.start` seems to have code to guard against that.

Reply:
Oh, we should definitely handle it, but if things get out-of-hand we might choose something easier over something "nicer."

Reply:
> ...It seems quite fair to use an AtomicBoolean and say "already started" even before the other start() returns. (I was mistaken and thought a duplicate start would just return immediately.)

Okay so you are suggesting that the first one saying "already started" really just means "start already called and in progress" as opposed to implying "already successfully started". So I am thinking of doing it in one of two ways:

- instead of a boolean have something that reflects 3 states: not_started, start_in_progress, successfully_started. This variable will reflect the state of this server object and guard against multiple (even concurrent) calls to `start`.

- keep the boolean and set it to true at the beginning of `start()` and set it to false if `start` does not succeed.

I would prefer to do it the second way as multiple concurrent start calls is not a use-case we should worry about so much.

Reply:
> keep the boolean and set it to true at the beginning of start() and set it to false if start does not succeed.

SGTM.

Reply:
done